2024-08-08 update : adding spatial to wwinference model.

cbernalz opened the pull request:
This PR is regarding the following issue : https://github.com/CDCgov/ww-inference-model/issues/44#issue-2445019536 .  The model that will be added into this will be coming from this issue : https://github.com/CDCgov/ww-inference-model/issues/25#issue-2409527383 .

cbernalz commented:
This is not ready yet, but I wanted to take input from both of you : @dylanhmorris and @kaitejohnson .

kaitejohnson commented:
> This is not ready yet, but I wanted to take input from both of you : @dylanhmorris and @kaitejohnson . Spatial components are in the model and using the testfile, I ran everything and things definitely change. There is a lot more variability, mainly in the forecast period. Rt take values that reach past 2, and site Rt's occasionally reach past 5.

Interesting! How does that compare to the ground truth R(t) values in each subpopulation? Might be worth making a plot of the estimated R(t)s and the known R(t)s that you simulated. 

Also, how tight are your priors around the ground truth values of $\phi$ and $\sigma_{\epsilon}$? To start, we might want to force those to be pretty tightly specified just to see if that allows us to reproduce the ground truth results (so cheating a little, but enforcing identifiability).

cbernalz commented:
> > This is not ready yet, but I wanted to take input from both of you : @dylanhmorris and @kaitejohnson . Spatial components are in the model and using the testfile, I ran everything and things definitely change. There is a lot more variability, mainly in the forecast period. Rt take values that reach past 2, and site Rt's occasionally reach past 5.
> 
> Interesting! How does that compare to the ground truth R(t) values in each subpopulation? Might be worth making a plot of the estimated R(t)s and the known R(t)s that you simulated.
> 
> Also, how tight are your priors around the ground truth values of ϕ and σ ϵ ? To start, we might want to force those to be pretty tightly specified just to see if that allows us to reproduce the ground truth results (so cheating a little, but enforcing identifiability).

Yes, I worked on changing these values and how big they are, but it seems like the thing that makes these values explode is using `sigma_generalized`, or the generalized variance formulation, from the normalization of the omega matrix.  When I removed this part from the model and did the traditional decomposition of $\Sigma$, I saw from the site Rt plots values are more realistic to what I believe the actual ones are.

I could have written the generalized variance into stan wrong, it might be good to look over that particular part @kaitejohnson  and @dylanhmorris on lines 236 in the wwinference.stan model.

dylanhmorris commented:
@cbernalz can you make some plots and share them in the other repo?

kaitejohnson commented:
> @cbernalz can you make some plots and share them in the other repo?

You could share them here because it only simulated data?

cbernalz commented:
@dylanhmorris and @kaitejohnson I am currently adjusting the `generate_simulation_data` function to use the generalized variance decomposition, I think this may have been the problem with the fitting.  I wanted to ask if we have the site Rt's with a generalized variance as c, what variance would I use for the aux site process.  Would I use this same c, so aux site deviations would have a $\text{Normal}(0,\text{scalingfactor}\cdot c)$?

kaitejohnson commented:
> @dylanhmorris and @kaitejohnson I am currently adjusting the `generate_simulation_data` function to use the generalized variance decomposition, I think this may have been the problem with the fitting. I wanted to ask if we have the site Rt's with a generalized variance as c , what variance would I use for the aux site process. Would I use this same c , so aux site deviations would have a Normal ( 0 , scalingfactor ⋅ c ) ?

Yep, I think that is the idea! So if `scalingfactor =1 ` on average each the auxiliary site subpopulation would have the same amount of deviation from the state R(t) as the (averaged across all sites) deviation from the state R(t) in the sites. But @dylanhmorris can correct me if this interpretation is wrong

cbernalz commented:
Great, @kaitejohnson and @dylanhmorris I figured out the problem was the strict prior placed on `sigma_generalized`.  The plots are on the ent repo issue.

kaitejohnson commented on R/generate_simulated_data.R:
```diff
@@ -579,7 +590,9 @@ generate_simulated_data <- function(r_in_weeks = # nolint
     hosp_data = hosp_data,
-    hosp_data_eval = hosp_data_eval
+    hosp_data_eval = hosp_data_eval,
+    rt_site_data = r_site, # temp?
```
@cbernalz This is where you are having the issue right, where when you call `example_data <- generate_simulated_data()` and then `example_data$rt` you get NULL? I would try the following: 
1. test everything within the generate simulated data function. Make sure `r_site` and `rt` are not NULL and are in fact a matrix and a vector respectively
2. try saving them as tibbles here (could also bind them to subpop spine and pivot them). I don't see why that would be the issue but am just trying to think what might be wrong here. 

Other thing is just to be absolutely sure you reinstalled the package before running the generate simulated data function but I am guessing you tried that already!

cbernalz replied:
Do you think it could be that reinstalling the package does not use the local version?  Because when I ran things just using the function and not with the package, I did get output for `rt_site_data` and `rt_global_data`.  When I don't do this and only use the package functions, I run each line of the `vignette_data.R`, `simulated_data$rt_global_data` and `simulated_data$rt_site_data` do not exist so this returns `NULL`.

kaitejohnson replied:
Hmm if you are using `remotes::install_local(force = TRUE)` I don't think so. Only thing I can think of is if you didn't reload the library (`library(wwinference)`)

kaitejohnson replied:
Just a question here, it seems that you were able to get all the package data saved. What was the fix?

dylanhmorris replied:
Was this resolved @cbernalz?

kaitejohnson commented:
Weirdly I am not able to see the full set of CI jobs (including the failure) for the `check-package`

kaitejohnson commented on R/data.R:
```diff
@@ -1,4 +1,4 @@
-#' Example wastewater dataset.
+#' Example wastewater dataset with site correlations from correlation func.
```
Maybe specify from an exponential correlation function with the specified distance matrix

kaitejohnson commented on R/data.R:
```diff
@@ -69,7 +109,44 @@
 #' @source vignette_data.R
 "hosp_data"
```
Should probably document this dataset and "hosp_data_eval" as being derived from the spatially correlated model.

kaitejohnson commented on R/data.R:
```diff
@@ -69,7 +109,44 @@
+#'   }
+#' @source vignette_data.R
+"hosp_data_ind"
+
```
I personally think this makes sense for your workflow to save both the spatially independent and dependent datasets as package data, but an alternative approach would just be to generate them directly in the analysis using the `generate_simulated_data()` function and tuning the parameters! And we might want to do that if we are going to vary some of the other parameters at some point (e.g. try different R(t) trajectories for example). Nothing to change now just a comment.

kaitejohnson commented on R/generate_simulated_data.R:
```diff
@@ -154,19 +154,19 @@ generate_simulated_data <- function(r_in_weeks = # nolint
                                         )
-                                      ),
-                                      phi = 25,
+                                      ) / 114.62984,
```
This can be a separate issue, but probably worth writing a function that normalizes the distance matrix within the `generate_simulated_data()` function, and just allows the "user" of this function to pass in their distance matrix. Since this right now isn't completely clear what you are doing.

dylanhmorris replied:
See also https://github.com/CDCgov/ww-inference-model/pull/56#discussion_r1722360146

kaitejohnson commented on R/generate_simulated_data.R:
```diff
@@ -381,7 +392,7 @@ generate_simulated_data <- function(r_in_weeks = # nolint
     log_r_site_aux <- construct_aux_rt(log_r_state_week,
       state_deviation_ar_coeff = phi_rt,
```
I think this would be slightly more readable if you specified the function arguments and passed them in e.g.
`log_r_site_aux <- construct_aux_rt(log_r_state_week = log_r_state_week) ` etc (so the same as you have for the second argument, but for all of them.

kaitejohnson replied:
I know this isn't done for everything else here, it will be once #29 is merged into the main codebase and into this....

dylanhmorris commented on R/generate_simulated_data.R:
```diff
@@ -57,8 +57,8 @@
 #' deviations
-#' @param sigma_eps Parameter for construction of covariance matrix of spatial
-#' epsilon
+#' @param sigma_generalized Parameter for construction of covariance
+#' matrix of spatial epsilon.  Generalized variance.
```
```suggestion
#' @param sigma_generalized  Generalized variance of the spatial epsilon
#' (determinant of the covariance matrix).
```

kaitejohnson commented on R/get_stan_data.R:
```diff
@@ -20,6 +20,8 @@
 #' @param compute_likelihood indicator variable telling stan whether or not to
 #' compute the likelihood, default = `1`
+#' @param dist_matrix Distance matrix for spatial correlation in distance
+#' correlation function.
```
Confirming this is not a dataframe but in fact a numeric matrix? Perhaps specify that it is  n_sites x n_sites

dylanhmorris commented on inst/stan/wwinference.stan:
```diff
@@ -117,7 +128,7 @@ parameters {
   real<lower=0, upper=1> autoreg_rt_site;
   real<lower=0, upper=1> autoreg_p_hosp;
-  matrix[n_subpops, n_weeks] error_site; // matrix of subpopulations
+  ////matrix[n_subpops, n_weeks] error_site; // matrix of subpopulations
```
Delete rather than comment out.
```suggestion
```

kaitejohnson replied:
Was going to suggest getting rid of this, and I think you should get rid of any "commented" code. But if it will help you make the modularity switch then I think its fine to leave it for now (though I think we would do that with the infrastucture you've written but passing in an identity matrix) instead of `sigma_mat`

kaitejohnson commented on R/wwinference.R:
```diff
@@ -68,7 +70,17 @@ wwinference <- function(ww_data,
                         generate_initial_values = TRUE,
-                        compiled_model = wwinference::compile_model()) {
+                        compiled_model = wwinference::compile_model(),
+                        dist_matrix = as.matrix(
```
I would recommend not having a default for this. Instead, I would pass it in right underneath the `ww_data` and `hosp_data` as a required argument to the function. Otherwise, the user could use this function without realizing they are supplying this distance matrix as input data.

kaitejohnson replied:
This will require you to add it to the vignette though and pass it to `wwinference(dist_matrix = etc)`.

For this PR, I would recommend just hard coding it into the vignette with the values you used to generate the spatially correlated data. In a separate issue, I would recommend adding the distance matrix used to generate the spatially correlated data as package data. That way, if you change it, it will be updated in the `wwinference.Rmd`

dylanhmorris replied:
> I would recommend not having a default for this. Instead, I would pass it in right underneath the `ww_data` and `hosp_data` as a required argument to the function. Otherwise, the user could use this function without realizing they are supplying this distance matrix as input data.

Agree, see https://github.com/CDCgov/ww-inference-model/pull/56#discussion_r1722360146

dylanhmorris replied:
Would strongly suggest default `NULL`, especially since we're also interested in cases where we infer a correlation structure but do not have a known distance matrix.

dylanhmorris reviewed (changes requested):
A few minor suggestions / comments, @cbernalz

kaitejohnson commented on data-raw/covid_pmfs.R:
```diff
@@ -7,20 +7,20 @@ params <- wwinference::get_params(
 # Put it all together
 generation_interval <- withr::with_seed(42, {
-  wwinference::simulate_double_censored_pmf(
+  wwinference:::simulate_double_censored_pmf(
```
```suggestion
  wwinference::simulate_double_censored_pmf(
```

kaitejohnson commented on data-raw/covid_pmfs.R:
```diff
@@ -7,20 +7,20 @@ params <- wwinference::get_params(
     max = params$gt_max, meanlog = params$mu_gi,
     sdlog = params$sigma_gi, fun_dist = rlnorm, n = 5e6
-  ) |> wwinference::drop_first_and_renormalize()
+  ) |> wwinference:::drop_first_and_renormalize()
```
```suggestion
  ) |> wwinference::drop_first_and_renormalize()
```

kaitejohnson commented on data-raw/covid_pmfs.R:
```diff
@@ -7,20 +7,20 @@ params <- wwinference::get_params(
 })
 
-inc <- wwinference::make_incubation_period_pmf(
+inc <- wwinference:::make_incubation_period_pmf(
```
```suggestion
inc <- wwinference::make_incubation_period_pmf(
```

kaitejohnson commented on data-raw/covid_pmfs.R:
```diff
@@ -7,20 +7,20 @@ params <- wwinference::get_params(
   params$backward_scale, params$backward_shape, params$r
 )
-sym_to_hosp <- wwinference::make_hospital_onset_delay_pmf(
+sym_to_hosp <- wwinference:::make_hospital_onset_delay_pmf(
```
```suggestion
sym_to_hosp <- wwinference::make_hospital_onset_delay_pmf(
```

kaitejohnson commented on data-raw/covid_pmfs.R:
```diff
@@ -7,20 +7,20 @@ params <- wwinference::get_params(
   params$neg_binom_size
 )
-inf_to_hosp <- wwinference::make_reporting_delay_pmf(inc, sym_to_hosp)
+inf_to_hosp <- wwinference:::make_reporting_delay_pmf(inc, sym_to_hosp)
```
```suggestion
inf_to_hosp <- wwinference::make_reporting_delay_pmf(inc, sym_to_hosp)
```

kaitejohnson replied:
I think these triple colors were introduced erroneously! They would error if you tried to run them but doesn't look like you did.

kaitejohnson commented on data-raw/test_data.R:
```diff
@@ -36,6 +36,17 @@ forecast_horizon <- 28
 generation_interval <- wwinference::generation_interval
 inf_to_hosp <- wwinference::inf_to_hosp
 
+dist_matrix <- as.matrix(
```
Related to my comment above and for a separate issue, I think you would want to just pull this in as package data corresponding to the spatially generated wastewater data and hospital admissions data, that way if you change it in the generate simulated data function and generate new package data it gets changed here too.

kaitejohnson commented on inst/stan/wwinference.stan:
```diff
@@ -87,6 +89,15 @@ data {
   real inf_feedback_prior_logmean;
   real<lower=0> inf_feedback_prior_logsd;
+
+  real log_phi_mu;
```
I would label these as priors e.g. `log_phi_mu_prior`. Also, we probably should rename either this parameter or our current `phi` parameter which is the dispersion in the negative binomial hospital admissions observation process.

kaitejohnson commented on inst/stan/wwinference.stan:
```diff
@@ -174,6 +193,18 @@ transformed parameters {
   vector[n_subpops] growth_site;
 
+  // Site spatial trans params--------------------------------------------------
+  real phi = exp(log_phi);
```
Ah I see, this doesn't conflict because we have called the hospital admissions observation process parameter `phi_h`. Would suggest making a separate issue.

kaitejohnson commented on inst/stan/wwinference.stan:
```diff
@@ -190,15 +221,45 @@ transformed parameters {
+    //                            1);
      //convert from weekly to daily
-     unadj_r_site_t = exp(to_row_vector(ind_m*(log_r_site_t_in_weeks)));
+     vector[n_weeks] log_r_site_t_in_weeks_vector = to_vector(combined_log_r_site_t_in_weeks[i, :]);
```
Would define this at the top of the transformed parameters block

kaitejohnson commented on inst/stan/wwinference.stan:
```diff
@@ -273,6 +334,14 @@ transformed parameters {
   // priors
+    // for spatial--------------------------------------------------------------
+    to_vector(non_cent_spatial_dev_ns_mat) ~ std_normal();
+    to_vector(norm_vec_aux_site) ~ std_normal();
```
I don't think you need the `to_vector()` because this is a vector already
```suggestion
    norm_vec_aux_site ~ std_normal();
```

kaitejohnson reviewed (commented):
@cbernalz This looks really really good! Great job putting this all together, and apologies for letting this PR get so big! 

My only high level suggestion that I think requires fixing in this PR is the comment about not putting the distance matrix in as default to the `wwinference()` wrapper function. And requiring the user specify that will require editing the vignette but it should be a pretty easy fix. The other stuff like moving the distance matrix to be package data as well can be done in another PR! I just want to make sure it's super easy for us to modify the distance matrix that we either pass in as a default to `generate_simulated_data()` or that we fit the model to in a simulation analysis.

cbernalz commented:
@dylanhmorris and @kaitejohnson I updated all of your comments and the biggest change is that I added a way to use either an independent correlation function (default if no `dist_matrix` is provided) and the exponential decay correlation function (if `dist_matrix` is provided).  Let me know what should be changed!

kaitejohnson commented on R/get_stan_data.R:
```diff
@@ -173,6 +174,20 @@ get_stan_data <- function(input_count_data,
+    # This dist_matrix will not be used, only needed for stan data specs.
+    dist_matrix <- matrix(
+      0,
+      nrow = subpop_data$n_subpops - 1,
```
The trick you are using here is that you are just passing in dummy data to stan that won't be used (but has the correct dimensions) right? Think this is a good approach!

kaitejohnson commented on inst/stan/wwinference.stan:
```diff
@@ -223,10 +229,15 @@ transformed parameters {
 
   // Site level spatial Rt------------------------------------------------------
-  non_norm_omega = exponential_decay_corr_func(dist_matrix, phi, l);
+  if (ind_corr_func == 1){
```
Essentially this is your spatial/non-spatial switch correct? And currently, the way they user would specify that would be whether or not they pass in a distance matrix? I think this is fine for now, I think we might want to think about making it a bit more explicit from the user side e.g. toggling spatial component/type of correlation function on and off regardless of the presence of the distance data but I think thats for another issue.

kaitejohnson reviewed (dismissed):
This looks good to me, sounds like you've also pretty much figured out how to do the switch on and off too which is great. See my one comment about how to specify that from the "user" side in the `wwinference()` function. But think thats for another PR.

Great work on this!! It's looking really nice. Excited to see it in action

dylanhmorris commented on inst/stan/wwinference.stan:
```diff
@@ -87,6 +91,16 @@ data {
+  real log_scaling_factor_mu_prior;
+  real log_scaling_factor_sd_prior;
+  matrix[n_subpops-1, n_subpops-1] dist_matrix;
+  real ind_corr_func;
```
```suggestion
  int<lower=0, upper=1> ind_corr_func;
```
This is a flag, yes? If so, should be an `int` equal to 0 (false) or 1 (true).

dylanhmorris commented on inst/stan/wwinference.stan:
```diff
@@ -190,15 +227,49 @@ transformed parameters {
   growth_site = initial_growth + eta_growth * sigma_growth; // site level growth rate
+
+  // Site level spatial Rt------------------------------------------------------
+  if (ind_corr_func == 1){
```
If you use an `int<lower=0, upper=1>`, can just call if
```suggestion
  if (ind_corr_func){
```

dylanhmorris replied:
See also https://mc-stan.org/docs/reference-manual/removals.html#real-values-in-conditionals

dylanhmorris commented on inst/stan/wwinference.stan:
```diff
@@ -190,15 +227,49 @@ transformed parameters {
+  // Site level spatial Rt------------------------------------------------------
+  if (ind_corr_func == 1){
+    // If no dist matrix given, use n_sites + 1 = n_subpops were all ind.
+    non_norm_omega = independence_corr_func(n_subpops - 1);
+  }
+  else {
+    non_norm_omega = exponential_decay_corr_func(norm_dist_matrix, phi, l);
+  }
+  norm_omega = matrix_normalization(non_norm_omega);
```
```suggestion
    non_norm_omega = independence_corr_func(n_subpops - 1);
    norm_omega = non_norm_omega;
  }
  else {
    non_norm_omega = exponential_decay_corr_func(norm_dist_matrix, phi, l);
    norm_omega = matrix_normalization(non_norm_omega);
  }
```

dylanhmorris replied:
Since the independence correlation function output is already determinant normalized, suggest skipping the normalization call when we use it.

dylanhmorris commented on inst/stan/wwinference.stan:
```diff
@@ -190,15 +227,49 @@ transformed parameters {
+  );
+  //----------------------------------------------------------------------------
+  // AUX site Rt----------------------------------------------------------------
+  log_r_auz_site_t_in_weeks = construct_aux_rt(
```
why `auz` rather than `aux`?

dylanhmorris commented on inst/stan/wwinference.stan:
```diff
@@ -190,15 +227,49 @@ transformed parameters {
-                                autoreg_rt_site, sigma_rt,
-                                to_vector(error_site[i]),
-                                1);
+    //log_r_site_t_in_weeks = ar1(log_r_mu_t_in_weeks,
+    //                            autoreg_rt_site, sigma_rt,
+    //                            to_vector(error_site[i]),
+    //                            1);
```
Remove commented code.
```suggestion
```

dylanhmorris commented on inst/stan/wwinference.stan:
```diff
@@ -281,7 +360,7 @@ model {
   autoreg_p_hosp ~ beta(autoreg_p_hosp_a, autoreg_p_hosp_b);
   log_r_mu_intercept ~ normal(r_logmean, r_logsd);
-  to_vector(error_site) ~ std_normal();
+  //to_vector(error_site) ~ std_normal();
```
```suggestion
```

dylanhmorris reviewed (changes requested):
This is looking very good @cbernalz. A few minor suggestions / fixes.

dylanhmorris commented on R/get_stan_data.R:
```diff
@@ -170,6 +175,20 @@ get_stan_data <- function(input_count_data,
+
+  # If dist_matrix null use independence correlation and update flag
+  if (is.null(dist_matrix)) {
+    ind_corr_func <- 1
```
```suggestion
    ind_corr_func <- 1L
```

dylanhmorris commented on R/get_stan_data.R:
```diff
@@ -170,6 +175,20 @@ get_stan_data <- function(input_count_data,
+      ncol = subpop_data$n_subpops - 1
+    )
+  } else {
+    ind_corr_func <- 0
```
```suggestion
    ind_corr_func <- 0L
```

dylanhmorris reviewed (dismissed):
LGTM, thanks @cbernalz 

A couple minor things I missed before about explicitly noting `1` and `0` as integers rather than floats (`1L`, `0L`). Can take or leave, and then merge.

dylanhmorris reviewed (approved):
Will let you merge, @cbernalz
